Scripts/download-portal-item-data.swift

```diff
@@ -36,55 +59,70 @@ func makeDataURL(portalURL: URL, itemIdentifier: String) -> URL {
-///   downloaded.
-///   - filename: The name of the downloaded file.
-/// - Returns: A new URL.
-func makeDownloadURL(downloadDirectoryURL: URL, filename: String) -> URL {
-    let subdirectories = [
-        "slpk": "Scene Layer Packages",
-        "mspk": "Mobile Scene Packages",
-        "tpk": "Tile Packages"
-    ]
-    var downloadURL = downloadDirectoryURL
-    if let subdirectory = subdirectories[(filename as NSString).pathExtension] {
-        downloadURL.appendPathComponent(subdirectory)
-    }
-    downloadURL.appendPathComponent(filename)
-    return downloadURL
+/// - Parameter url: The url of a ZIP archive.
+/// - Returns: The file name.
+func nameOfFileInArchive(at url: URL) throws -> String {
+    let outputPipe = Pipe()
+    let process = Process()
+    process.executableURL = URL(fileURLWithPath: "/usr/bin/zipinfo", isDirectory: false)
+    process.arguments = ["-1", url.path]
+    process.standardOutput = outputPipe
+    try process.run()
+    process.waitUntilExit()
+    
+    let filenameData = outputPipe.fileHandleForReading.readDataToEndOfFile()
+    return String(data: filenameData, encoding: .utf8)!.trimmingCharacters(in: .whitespacesAndNewlines)
 }
 
-extension FileManager {
-    /// Indicates whether a file or directory exists at the given URL.
-    ///
-    /// - Parameter url: A URL.
-    /// - Returns: `true` if a file or directory at the given path exists,
-    /// otherwise `false`.
-    func fileExists(at url: URL) -> Bool {
-        return fileExists(atPath: url.path)
-    }
-}
-
-/// Downloads the file at the given URL to the given URL.
+/// Uncompresses the data in the archive at the source URL into the destination
+/// URL.
 ///
 /// - Parameters:
-///   - sourceURL: The URL of the file to download.
-///   - destinationURL: The URL to which the file should be downloaded.
-func downloadFile(at sourceURL: URL, to destinationURL: URL, completion: @escaping (Error?) -> Void) {
-    let downloadTask = URLSession.shared.downloadTask(with: sourceURL) { (urlOrNil, responseOrNil, errorOrNil) in
-        if let fileURL = urlOrNil {
+///   - sourceURL: The URL of a ZIP archive.
+///   - destinationURL: The URL at which to uncompress the archive.
+func uncompressArchive(at sourceURL: URL, to destinationURL: URL) throws {
+    FileManager.default.createFile(atPath: destinationURL.path, contents: nil)
+    let fileHandle = try FileHandle(forWritingTo: destinationURL)
+    let process = Process()
+    process.executableURL = URL(fileURLWithPath: "/usr/bin/unzip", isDirectory: false)
+    process.arguments = ["-p", sourceURL.path]
+    process.standardOutput = fileHandle
+    try process.run()
+    process.waitUntilExit()
+}
+
+func downloadFile(at sourceURL: URL, destinationURLProvider: URLProvider, completion: @escaping (Result<URL, Error>) -> Void) {
+    let downloadTask = URLSession.shared.downloadTask(with: sourceURL) { (temporaryURL, response, error) in
```

Comment:
if you are downloading something off a portal, you should use `AGSRequestOperation` so that it takes care of all authentication for you. You can give it an output url which will allow it to save the download somewhere on disk instead of in memory.

Reply:
It was a conscious decision to not have `ArcGIS` as a dependency of this script. All of the portal items downloaded by this script are public.